corehq/form_processor/backends/couch/processor.py

```diff
@@ -147,39 +147,45 @@ def get_cases_from_forms(case_db, xforms):
         return touched_cases
 
     @staticmethod
-    def hard_rebuild_case(domain, case_id, detail, save=True):
-        try:
-            case = CommCareCase.get(case_id)
-            assert case.domain == domain
-            found = True
-        except CaseNotFound:
+    def hard_rebuild_case(domain, case_id, detail, save=True, lock=True):
+        case, lock_obj = FormProcessorCouch.get_case_with_lock(case_id, lock=lock)
+        found = bool(case)
+        if not found:
             case = CommCareCase()
             case.case_id = case_id
             case.domain = domain
-            found = False
+            if lock:
+                lock_obj = CommCareCase.get_obj_lock_by_id(case_id)
 
-        forms = FormProcessorCouch.get_case_forms(case_id)
-        filtered_forms = [f for f in forms if f.is_normal]
-        sorted_forms = sorted(filtered_forms, key=lambda f: f.received_on)
-
-        actions = _get_actions_from_forms(domain, sorted_forms, case_id)
-
-        if not found and case.domain is None:
-            case.domain = domain
-
-        rebuild_case_from_actions(case, actions)
-        # todo: should this move to case.rebuild?
-        if not case.xform_ids:
-            if not found:
-                return None
-            # there were no more forms. 'delete' the case
-            case.doc_type = 'CommCareCase-Deleted'
-
-        # add a "rebuild" action
-        case.actions.append(_rebuild_action())
-        if save:
-            case.save()
-        return case
+        assert case.domain == domain
```

Comment:
This should move into the `try` block. Otherwise you could end up with a lock that never gets released in case of assertion failure. Sorry I didn't catch this before because it wasn't clear to me what was happening with the `lock_obj` on line 151.

Nit: the error message of this assertion (if it fails) will not be very informative. Suggested change:
```py
assert case.domain == domain, (case.domain, domain)
```

Reply:
Ah, yes, forgetting that if we got a lock initially then this won't release it. I'd thought there was a reason I'd had it that way initially but couldn't remember.

Reply:
6222ce7